src/modules/fw_auto_trim/FwAutoTrim.cpp

```diff
@@ -0,0 +1,308 @@
+			_landed = vehicle_land_detected.landed;
+		}
+	}
+
+
+	vehicle_torque_setpoint_s vehicle_torque_setpoint;
+
+	if (!_vehicle_torque_setpoint_sub.update(&vehicle_torque_setpoint)
+	    || (vehicle_torque_setpoint.timestamp == _timestamp_last)) {
+		return;
+	}
+
+	// check for parameter updates
+	if (_parameter_update_sub.updated()) {
+		// clear update
+		parameter_update_s pupdate;
+		_parameter_update_sub.copy(&pupdate);
+
+		// update parameters from storage
+		updateParams();
+	}
+
+	perf_begin(_cycle_perf);
+
+	if (_vehicle_status_sub.updated()) {
+		vehicle_status_s vehicle_status;
+
+		if (_vehicle_status_sub.copy(&vehicle_status)) {
+			_armed = (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED);
+			_fixed_wing = (vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING);
+			_tailsitter = vehicle_status.is_vtol_tailsitter;
+		}
+	}
+
+	if (_airspeed_validated_sub.updated()) {
+		airspeed_validated_s airspeed_validated;
+
+		if (_airspeed_validated_sub.copy(&airspeed_validated)) {
+			_airspeed = airspeed_validated.true_airspeed_m_s;
+		}
+	}
+
+	if (_vehicle_attitude_sub.updated()) {
+		vehicle_attitude_s vehicle_attitude;
+
+		if (_vehicle_attitude_sub.copy(&vehicle_attitude)) {
+			const Vector3f earth_z_in_body_frame = matrix::Quatf(vehicle_attitude.q).rotateVectorInverse(Vector3f(0.f, 0.f, 1.f));
+			_cos_tilt = _tailsitter ? earth_z_in_body_frame(0) : earth_z_in_body_frame(2);
+		}
+	}
+
+	const float dt = (vehicle_torque_setpoint.timestamp - _timestamp_last) * 1e-6f;
+	_timestamp_last = vehicle_torque_setpoint.timestamp;
+
+	const hrt_abstime now = hrt_absolute_time();
+	const Vector3f torque{vehicle_torque_setpoint.xyz};
+	state state_prev = _state;
+
+	const bool run_auto_trim = _fixed_wing
+				   && _armed
+				   && !_landed
+				   && (dt > 0.001f) && (dt < 1.f)
+				   && torque.isAllFinite()
+				   && _airspeed >= _param_fw_airspd_min.get()
+				   && _airspeed <= _param_fw_airspd_max.get()
+				   && _cos_tilt > cosf(math::radians(_kTiltMaxDeg));
+
+	if (run_auto_trim) {
+		switch (_state) {
+		default:
+
+		// fallthrough
+		case state::idle:
+			_trim_estimate.reset();
+			_trim_test.reset();
+			_state = state::sampling;
+			_state_start_time = now;
+			break;
+
+		case state::sampling:
+			_trim_estimate.update(torque);
+
+			if ((now - _state_start_time) > 5_s) {
+				_state = state::sampling_test;
+				_state_start_time = now;
+			}
+
+			break;
+
+		case state::sampling_test:
+			_trim_test.update(torque);
+
+			if ((now - _state_start_time) > 2_s) {
+				_state = state::verification;
+				_state_start_time = now;
+			}
+
+			break;
+
+		case state::verification: {
+				// Perform a statistical test between the estimated and test data
+				const float gate = 5.f;
+				const Vector3f innovation = _trim_test.mean() - _trim_estimate.mean();
+				const Vector3f innovation_variance = _trim_test.variance() + _trim_test.variance();
```

Comment:
The innovation variance is twice the test variance? Or do you have a typo here?

Reply:
ouch, typo